components/camel-reactive-streams/src/main/java/org/apache/camel/component/reactive/streams/engine/CamelPublisher.java

```diff
@@ -0,0 +1,105 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.camel.component.reactive.streams.engine;
+
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.apache.camel.CamelContext;
+import org.apache.camel.Exchange;
+import org.apache.camel.component.reactive.streams.ReactiveStreamsBackpressureStrategy;
+import org.apache.camel.component.reactive.streams.ReactiveStreamsComponent;
+import org.apache.camel.component.reactive.streams.api.DispatchCallback;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The Camel publisher. It forwards Camel exchanges to external reactive-streams subscribers.
+ */
+public class CamelPublisher implements Publisher<StreamPayload<Exchange>>, AutoCloseable {
+
+    private static final Logger LOG = LoggerFactory.getLogger(CamelPublisher.class);
+
+    private ExecutorService workerPool;
+
+    private String name;
+
+    private ReactiveStreamsBackpressureStrategy backpressureStrategy;
+
+    private List<CamelSubscription> subscriptions = new CopyOnWriteArrayList<>();
+
+    public CamelPublisher(ExecutorService workerPool, CamelContext context, String name) {
+        this.workerPool = workerPool;
+        this.backpressureStrategy = ((ReactiveStreamsComponent) context.getComponent("reactive-streams")).getBackpressureStrategy();
+        this.name = name;
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super StreamPayload<Exchange>> subscriber) {
+        Objects.requireNonNull(subscriber, "subscriber must not be null");
+        CamelSubscription sub = new CamelSubscription(workerPool, this, this.backpressureStrategy, subscriber);
+        this.subscriptions.add(sub);
+        subscriber.onSubscribe(sub);
+    }
+
+    public void unsubscribe(CamelSubscription subscription) {
+        subscriptions.remove(subscription);
+    }
+
+    public void publish(StreamPayload<Exchange> data) {
+        // freeze the subscriptions
+        List<CamelSubscription> subs = new LinkedList<>(subscriptions);
+        DispatchCallback<Exchange> originalCallback = data.getCallback();
+        if (originalCallback != null && subs.size() > 0) {
+            // Notify processing once if multiple subscribers are present
+            AtomicInteger counter = new AtomicInteger(0);
+            AtomicReference<Throwable> thrown = new AtomicReference<>(null);
+            data = new StreamPayload<>(data.getItem(), (ex, error) -> {
+                int status = counter.incrementAndGet();
+                thrown.compareAndSet(null, error);
+                if (status == subs.size()) {
```

Comment:
Small optimization, it may also be implemented as:

```java
AtomicInteger counter = new AtomicInteger(subs.size());
...
if (counter.decrementAndGet() == 0) {
    ...
}
```